alloptions-limit.php

```diff
@@ -32,45 +34,50 @@ function wpcom_vip_sanity_check_alloptions() {
 
 	// Cache miss
 	if ( false === $alloptions_size ) {
-		// phpcs:ignore WordPress.PHP.DiscouragedPHPFunctions.serialize_serialize
-		$alloptions      = serialize( wp_load_alloptions() );
+		$alloptions      = maybe_serialize( wp_load_alloptions() );
 		$alloptions_size = strlen( $alloptions );
 
 		wp_cache_add( 'alloptions_size', $alloptions_size, '', 60 );
 	}
 
-	$blocked = $alloptions_size > $alloptions_size_die;
-	$warning = $alloptions_size > $alloptions_size_warn;
+	$warning        = $alloptions_size > $alloptions_size_warn;
+	$maybe_blocked  = $alloptions_size > $alloptions_size_die;
+	$really_blocked = false;
 
-	// If it's at least over the warning threshold (will also run when blocked), notify
-	if ( $warning ) {
-		if ( $blocked ) {
-			add_filter( 'alloptions_overrule_ack', '__return_true' );
-		}
-		// NOTE - This function has built-in rate limiting so it's ok to call on every request
-		wpcom_vip_sanity_check_alloptions_notify( $alloptions_size, $blocked );
+	$alloptions_size_compressed = 0;
+
+	if ( ! $warning ) {
+		return;
 	}
 
-	// Will exit with a 503
-	if ( $blocked ) {
-		wpcom_vip_sanity_check_alloptions_die( $alloptions );
+	if ( $maybe_blocked ) {
+		// It's likely at this point the site is already experiencing performance degradation.
+		// We're using gzdeflate here because pecl-memcache uses Zlib compression for large values.
+		// See https://github.com/websupport-sk/pecl-memcache/blob/e014963c1360d764e3678e91fb73d03fc64458f7/src/memcache_pool.c#L303-L354
+		$alloptions_size_compressed = wp_cache_get( 'alloptions_size_compressed' );
+		if ( ! $alloptions_size_compressed ) {
+			$alloptions_size_deflated = gzdeflate( maybe_serialize( wp_load_alloptions() ) );
+			if ( $alloptions_size_deflated ) {
+				$alloptions_size_compressed = strlen( $alloptions_size_deflated );
+				wp_cache_add( 'alloptions_size_compressed', $alloptions_size_compressed, '', 60 );
+			}
```

Comment:
In case of a fail during `gzdeflate()` (not sure when that would be), I think maybe better to still cache so it doesn't keep trying, and then just assume something is wrong (but not "time to block site" wrong):

```
$alloptions_size_deflated = gzdeflate( maybe_serialize( wp_load_alloptions() ) );
$alloptions_size_compressed = $alloptions_size_deflated !== false ? strlen( $alloptions_size_deflated ) : VIP_ALLOPTIONS_ERROR_THRESHOLD - 1;
wp_cache_add( 'alloptions_size_compressed', $alloptions_size_compressed, '', 60 );
```